### PR TITLE
addressing a NPE in the compiler due to empty warnStack

### PR DESCRIPTION
**Pull request by vicente-romero-oracle**
Fixing a compiler crash plus other review comments from Maurizio

---------
### Progress
- [x] Change must not contain extraneous whitespace

### Reviewers
 * [Maurizio Cimadamore](https://openjdk.java.net/census#mcimadamore) (@mcimadamore - Committer)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/valhalla pull/663/head:pull/663` \
`$ git checkout pull/663`

Update a local copy of the PR: \
`$ git checkout pull/663` \
`$ git pull https://git.openjdk.java.net/valhalla pull/663/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 663`

View PR using the GUI difftool: \
`$ git pr show -t 663`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/valhalla/pull/663.diff">https://git.openjdk.java.net/valhalla/pull/663.diff</a>

</details>

### CODE REVIEW

File: src/jdk.compiler/share/classes/com/sun/tools/javac/code/Types.java [hunk 1]
```diff
@@ -1672,8 +1685,18 @@ public boolean containsType(Type t, Type s) {
             public Boolean visitType(Type t, Type s) {
                 if (s.isPartial())
                     return containedBy(s, t);
-                else
-                    return isSameType(t, s);
+                else {
+                    boolean result = isSameType(t, s);
+                    // warnStack.head is != null if we are checking for an assignment, in other cases we should be strict
+                    // the order in the condition below matters
+                    if (warnStack.head != null && allowUniversalTVars && !result) {
```
Comment (vicente-romero-oracle) [inline src/jdk.compiler/share/classes/com/sun/tools/javac/code/Types.java:1692]: using the emptiness of the warnStack as the way to know if we are dealing with an assignment or not seems a bit fragile, another option could be to pass down a boolean all the way down from all invocations, what do you think?

Comment (mcimadamore) [reply src/jdk.compiler/share/classes/com/sun/tools/javac/code/Types.java:1692]: Do you mean, e.g. inside overload resolution, how do we distinguish between when we can apply the "loose" check, and when we need the strict check?

If so, my feeling here is that as we have isSubtype vs. isSubtypeUnchecked we also need isSameTypes vs. isSameTypesUnchecked and a subtyping routine should call the "correct" isSameTypes visitor/routine.

**Review by mcimadamore — COMMENTED** [review]
My suggestion is to fix the crash first, and then file another PR for adding support for nested unchecked null warnings, as that is probably a biggie feature.

File: src/jdk.compiler/share/classes/com/sun/tools/javac/code/Types.java [hunk 2]
```diff
@@ -1727,8 +1750,12 @@ public Boolean visitUndetVar(UndetVar t, Type s) {
             public Boolean visitTypeVar(TypeVar t, Type s) {
                 if (s.hasTag(TYPEVAR)) {
                     TypeVar other = (TypeVar)s;
-                    if (allowUniversalTVars && t.isValueProjection() != other.isValueProjection() && t.tsym == other.tsym)
+                    if (allowUniversalTVars && t.isValueProjection() != other.isValueProjection() && t.tsym == other.tsym) {
+                        if (warnStack.head != null) {
```
[the remark this answers is not in the document]
Comment (vicente-romero-oracle) [reply src/jdk.compiler/share/classes/com/sun/tools/javac/code/Types.java:1754]: correct

**Comment by vicente-romero-oracle** [comment]
> My suggestion is to fix the crash first, and then file another PR for adding support for nested unchecked null warnings, as that is probably a biggie feature.

sounds good I will do that

**Review by mcimadamore — APPROVED** [review]
The fix makes sense, and seems consistent with what we do with isCastable.

**Pull request closed without merging**